Statefulset should wait for new replicas when removing .start.ordinal

soltysh opened the pull request:
#### What type of PR is this?
/kind flake
/sig apps
/sig testing

#### What this PR does / why we need it:
We've noticed this problem in OpenShift CI flaking rather frequently. The problematic test is the one where we create StatefulSet with `.start.ordinal` set, after getting the StatefulSet ready we remove `.start.ordinal` and wait for two new pods to show up. But the test was written in such a way that it first checked for 2 ready replicas and when it got them it proceeded to checking if the correct pods appeared, which in some cases didn't happen.  In those flaky cases the first wait for 2 ready replicas caught the initial 2 pods sometimes, causing the wait to be close to zero, which didn't give time for the controller to react and roll the new pods. 

This proposed fix changes the ordering, such that we will first wait for the 2 new pods to appear and only after that make sure that the StatefulSet reports full readiness.

#### Which issue(s) this PR fixes:
Described in https://issues.redhat.com//browse/TRT-1179

#### Special notes for your reviewer:
/assign @atiratree @bertinatto 

#### Does this PR introduce a user-facing change?
```release-note
Statefulset should wait for new replicas in tests when removing .start.ordinal
```

soltysh committed:
Statefulset should wait for new replicas when removing .start.ordinal

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1595,19 +1601,32 @@ var _ = SIGDescribe("StatefulSet", func() {
+					if intSet.Status.ObservedGeneration < ss.Generation {
+						return false, nil
+					}
+					if err := expectPodNames(pl, []string{"ss-0", "ss-1", "ss-2"}); err != nil {
```
The fix looks good to me, but why do we wait for 3 pods? Aren't we expecting two replicas? `"ss-0"` and `"ss-1"`? It might skip this state since the poll period is 10s.

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1579,7 +1584,8 @@ var _ = SIGDescribe("StatefulSet", func() {
 
```
I can see we are doing `waitForStatus` in the other tests, but isn't there a chance for a flake as well? Shouldn't we try to confirm we have the correct set of replicas before we check they are ready?

soltysh replied:
Good catch, yes we should definitely do a similar wait as in the other, lemme fix it...

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1515,7 +1516,8 @@ var _ = SIGDescribe("StatefulSet", func() {
 
```
this test doesn't even `waitForStatus`, even though it probably never flakes since the number of replicas changes. It might be worth to use the same approach here as well.

soltysh replied:
ack

atiratree commented:
LGTM and the test pass locally

unfortunately we do not seem to run these tests in CI as all `Feature` labeled tests are skipped

https://github.com/kubernetes/test-infra/blob/9643c065212d4fb1698bfbe2302176d69028f1fb/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L255

unless they are especially enabled in pull-kubernetes-e2e-gce-cos-alpha-features https://github.com/kubernetes/test-infra/blob/9643c065212d4fb1698bfbe2302176d69028f1fb/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L363 which these are not

Can we remove the `[Feature:StatefulSetStartOrdinal]` label for a beta feature? Not sure what is the policy there...

soltysh committed:
Drop [Feature: tag to make sure this tests runs on pre-submits

In v1.27 StatefulSetStartOrdinal became beta, which makes it on by
default, but we forgot to turn these tests on along with it. This makes
these tests run always.

soltysh commented:
@pwschuurman fyi, especially the 2nd commit which makes these test run always in our CI.

soltysh commented:
> Can we remove the `[Feature:StatefulSetStartOrdinal]` label for a beta feature? Not sure what is the policy there...

Yes, those should've been removed when we promoted this feature to beta, I've added 2nd commit doing just that, and also pinged Peter, responsible for that feature.

atiratree commented:
seeing all the tests pass in E2Es
/lgtm
/retest
